block-producer/src/block_builder/account.rs

```diff
@@ -0,0 +1,153 @@
+use miden_air::ExecutionOptions;
+use miden_objects::{
+    accounts::AccountId,
+    assembly::Assembler,
+    crypto::merkle::{MerklePath, MerkleStore},
+    Digest, Felt,
+};
+use miden_stdlib::StdLibrary;
+use miden_vm::{
+    crypto::MerkleError, execute, AdviceInputs, DefaultHost, ExecutionError, MemAdviceProvider,
+    Program, StackInputs,
+};
+use thiserror::Error;
+
+#[derive(Error, Debug)]
+pub enum AccountRootUpdateError {
+    #[error("Received invalid merkle path")]
+    InvalidMerklePaths(MerkleError),
+    #[error("program execution failed")]
+    ProgramExecutionFailed(ExecutionError),
+    #[error("invalid return value on stack (not a hash)")]
+    InvalidRootReturned,
+}
+
+/// Stack inputs:
+/// [num_accounts_updated,
+///  ACCOUNT_ROOT,
+///  NEW_ACCOUNT_HASH_0, account_id_0, ... , NEW_ACCOUNT_HASH_n, account_id_n]
+const ACCOUNT_UPDATE_ROOT_MASM: &str = "
+use.std::collections::smt64
+
+begin
+    push.1
+    while.true
+        # stack: [counter, ROOT_0, ..., NEW_ACCOUNT_HASH_i, account_id_i , ...]
+
+        # Move counter down for next iteration
+        movdn.9
+        # => [ROOT_i, NEW_ACCOUNT_HASH_i, account_id_i, counter, ...]
+
+        # Move root down (for smt64.set)
+        movdn.8 movdn.8 movdn.8 movdn.8
+        # => [NEW_ACCOUNT_HASH_i, account_id_i, ROOT_i, counter, ...]
+
+        # set new value in SMT
+        smt64.set dropw
+        # => [ROOT_{i+1}, counter, ...]
+
+        # loop counter
+        movup.4 sub.1 dup neq.0
+        # => [0 or 1, counter-1, ROOT_{i+1}, ...]
+    end
+
+    drop
+    # => [ROOT_{n-1}]
+end
+";
```

Comment:
This is fine for now. Eventually, this will become a "block kernel" - a program which takes old block header, and a commitment to the batches as an input and outputs the new block header (at that point, it'll probably live in `miden-lib` crate).

When we implement nullifier handling, we might want to combine both programs into a single one which will serve as a very early prototype of the block kernel.